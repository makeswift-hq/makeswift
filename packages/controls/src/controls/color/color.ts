import { match } from 'ts-pattern'
import { z } from 'zod'

import { safeParse, type ParseResult } from '../../lib/zod'

import { ControlDataTypeKey, Schema } from '../../common'
import { type CopyContext } from '../../context'
import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'

import { ControlDefinition, serialize, type SchemaType } from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'

type Config = z.infer<typeof Definition.schema.relaxed.config>

type SchemaByDefaultValue<D extends Config['defaultValue']> =
  undefined extends D
    ? typeof Definition.schema.relaxed
    : typeof Definition.schema.strict

type Schema<C extends Config> = SchemaByDefaultValue<C['defaultValue']>
type DataType<C extends Config> = z.infer<Schema<C>['data']>
type ValueType<C extends Config> = z.infer<Schema<C>['value']>
type ResolvedValueType<C extends Config> = z.infer<Schema<C>['resolvedValue']>

type ReturnedSchemaType<C extends Config> = {
  definition: typeof Definition.schema.relaxed.definition
  type: typeof Definition.schema.relaxed.type
  data: SchemaType<DataType<C>>
  value: SchemaType<ValueType<C>>
  resolvedValue: SchemaType<ResolvedValueType<C>>
}

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  private static readonly v1DataType = 'color::v1' as const
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const

  static readonly type = 'makeswift::controls::color' as const

  static get schema() {
    const type = z.literal(this.type)
    const version = z.literal(1).optional()
    const value = Schema.colorData
    const resolvedSwatchValue = Schema.resolvedColorData

    const schemas = <R>(resolvedValue: SchemaType<R>) => {
      const data = value.merge(
        z.object({
          [ControlDataTypeKey]: z.literal(this.v1DataType).optional(),
        }),
      )

      const config = z.object({
        label: z.string().optional(),
        labelOrientation: z
          .union([z.literal('horizontal'), z.literal('vertical')])
          .optional(),
        defaultValue: resolvedValue,
        hideAlpha: z.boolean().optional(),
      })

      const definition = z.object({
        type,
        config,
        version,
      })

      return {
        type,
        value,
        resolvedSwatchValue,
        resolvedValue,
        data,
        config,
        version,
        definition,
      }
    }

    return {
      version,
      relaxed: schemas(z.string().optional()),
      strict: schemas(z.string()),
    }
  }

  static deserialize(data: DeserializedRecord): ColorDefinition {
    if (data.type !== Definition.type)
      throw new Error(
        `Color deserialization: expected '${Definition.type}', got '${data.type}'`,
      )

    const { config, version } = Definition.schema.relaxed.definition.parse(data)
    return new ColorDefinition(config, version)
  }

  constructor(
    config: C,
    readonly version: z.infer<typeof Definition.schema.version>,
  ) {
    super(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema(): ReturnedSchemaType<C> {
    return Definition.schema.relaxed
  }

  get dataSchema() {
    return (
      (this.config.defaultValue === undefined
        ? Definition.schema.relaxed
        : Definition.schema.strict) as Schema<C>
    ).data
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.dataSchema, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, ({ swatchId, alpha }) => ({
        swatchId,
        alpha,
      }))
      .otherwise((value) => value)
  }

  toData(value: ValueType<C>): DataType<C> {
    return match(this.version)
      .with(1, () => ({
        ...Definition.dataSignature.v1,
        ...value,
      }))
      .with(undefined, () => value)
      .exhaustive()
  }

  copyData(
    data: DataType<C> | undefined,
    { replacementContext }: CopyContext,
  ): DataType<C> | undefined {
    if (data == null) return data

    const replaceSwatchId = (swatchId: string) =>
      replacementContext.swatchIds.get(swatchId) ?? swatchId

    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, (val) => ({
        ...val,
        swatchId: replaceSwatchId(val.swatchId),
      }))
      .otherwise((val) => ({
        ...val,
        swatchId: replaceSwatchId(val.swatchId),
      }))
  }

  createInstance(sendMessage: SendMessage<any>) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
      version: this.version,
    })
  }
}

export class ColorDefinition<C extends Config = Config> extends Definition<C> {}

type UserConfig<D extends Config['defaultValue']> = Config & {
  defaultValue?: D
}

type NormedConfig<D extends Config['defaultValue']> = z.infer<
  SchemaByDefaultValue<D>['config']
>

export function Color<D extends Config['defaultValue']>(
  config?: UserConfig<D>,
): ColorDefinition<NormedConfig<D>> {
  return new ColorDefinition<NormedConfig<D>>(config ?? {}, 1)
}
