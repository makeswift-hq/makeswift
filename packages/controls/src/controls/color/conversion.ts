import parseColor from 'color'

import { type Swatch } from '../../resources'

export function swatchToColorString(
  swatch: Swatch | null,
  alpha: number,
  defaultValue?: string,
): string | undefined {
  if (swatch == null) {
    return defaultValue === undefined
      ? undefined
      : safeParseColor(defaultValue).rgb().string()
  }

  return parseColor({
    h: swatch.hue,
    s: swatch.saturation,
    l: swatch.lightness,
  })
    .alpha(alpha)
    .rgb()
    .string()
}

function safeParseColor(value: string) {
  try {
    return parseColor(value)
  } catch {
    return parseColor()
  }
}
