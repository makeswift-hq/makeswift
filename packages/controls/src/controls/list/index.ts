export * from './list'
export * from './list-control'
