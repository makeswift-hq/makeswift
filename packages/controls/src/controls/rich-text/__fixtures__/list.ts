import { type RichTextDTO } from '../dto'

export const list: RichTextDTO = {
  object: 'value',
  annotations: undefined,
  data: undefined,
  document: {
    data: undefined,
    nodes: [
      {
        data: {},
        type: 'unordered-list',
        nodes: [
          {
            data: {},
            type: 'list-item',
            nodes: [
              {
                data: {},
                type: 'list-item-child',
                nodes: [
                  {
                    text: 'Point1',
                    marks: [
                      {
                        data: {
                          value: {
                            id: 'VHlwb2dyYXBoeTpiZjFhNzNlNi0yMGNkLTQwZjEtOThmMC02MDc5OWU2ZWJmZmE=',
                            style: [],
                          },
                        },
                        type: 'typography',
                        object: 'mark',
                      },
                    ],
                    object: 'text',
                  },
                ],
                object: 'block',
              },
            ],
            object: 'block',
          },
          {
            data: {},
            type: 'list-item',
            nodes: [
              {
                data: {},
                type: 'list-item-child',
                nodes: [
                  {
                    text: 'POint 2',
                    marks: [
                      {
                        data: {
                          value: {
                            id: 'VHlwb2dyYXBoeTpiZjFhNzNlNi0yMGNkLTQwZjEtOThmMC02MDc5OWU2ZWJmZmE=',
                            style: [],
                          },
                        },
                        type: 'typography',
                        object: 'mark',
                      },
                    ],
                    object: 'text',
                  },
                ],
                object: 'block',
              },
            ],
            object: 'block',
          },
          {
            data: {},
            type: 'list-item',
            nodes: [
              {
                data: {},
                type: 'list-item-child',
                nodes: [
                  {
                    text: 'POint 3',
                    marks: [
                      {
                        data: {
                          value: {
                            id: 'VHlwb2dyYXBoeTpiZjFhNzNlNi0yMGNkLTQwZjEtOThmMC02MDc5OWU2ZWJmZmE=',
                            style: [],
                          },
                        },
                        type: 'typography',
                        object: 'mark',
                      },
                    ],
                    object: 'text',
                  },
                ],
                object: 'block',
              },
            ],
            object: 'block',
          },
        ],
        object: 'block',
      },
    ],
    object: 'document',
  },
  selection: undefined,
}
