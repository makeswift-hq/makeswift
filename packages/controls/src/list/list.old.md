// import {
//   NumberControlDefinition,
//   TextInputControlDefinition,
//   TextAreaControlDefinition,
//   Data,
//   ComboboxControlData,
//   ComboboxControlDefinition,
//   IconRadioGroupControlData,
//   IconRadioGroupControlDefinition,
//   ImageControlData,
//   ImageControlDefinition,
//   LinkControlData,
//   LinkControlDefinition,
// } from '@makeswift/controls'

// import {
//   AnyPropController,
//   createPropController,
//   PropControllerMessage,
//   Send,
// } from '../../../runtime/src/prop-controllers/instances'

// import { PropController } from '../../../runtime/src/prop-controllers/base'
// import {
//   CopyContext,
//   MergeTranslatableDataContext,
// } from '../../../runtime/src/state/react-page'

// import {
//   ControlDefinition,
//   ControlDefinitionData,
//   getTranslatableData,
//   mergeTranslatedData,
// } from '../../../runtime/src/controls/control'

// import { copy as controlCopy } from '../../../runtime/src/controls/control'
// import {
//   getElementChildren,
//   getFileIds,
//   getPageIds,
//   getSwatchIds,
//   getTypographyIds,
// } from '../../../runtime/src/prop-controllers/introspection'
// import {
//   SelectControlData,
//   SelectControlDefinition,
// } from '../../../runtime/src/controls/select'
// import { ShapeControlDefinition } from '../../../runtime/src/controls/shape'
// import {
//   RichTextControlData,
//   RichTextControlDefinition,
// } from '../../../runtime/src/controls/rich-text'
// import {
//   RichTextV2ControlData,
//   RichTextV2ControlDefinition,
// } from '../../../runtime/src/controls/rich-text-v2'
// import {
//   StyleControlData,
//   StyleControlDefinition,
// } from '../../../runtime/src/controls/style'
// import {
//   StyleV2ControlData,
//   StyleV2ControlDefinition,
// } from '../../../runtime/src/controls/style-v2'
// import {
//   TypographyControlData,
//   TypographyControlDefinition,
// } from '../../../runtime/src/controls/typography'
// import { IndexSignatureHack } from '../../../runtime/src/utils/index-signature-hack'

// export type GetItemLabelControlData<T extends ControlDefinition> =
//   T extends NumberControlDefinition
//     ? number
//     : T extends TextInputControlDefinition
//       ? string
//       : T extends TextAreaControlDefinition
//         ? string
//         : T extends SelectControlDefinition
//           ? SelectControlData<T>
//           : T extends IconRadioGroupControlDefinition
//             ? IconRadioGroupControlData<T>
//             : T extends ImageControlDefinition
//               ? ImageControlData
//               : T extends ComboboxControlDefinition
//                 ? ComboboxControlData<T>
//                 : T extends ShapeControlDefinition
//                   ? GetItemLabelShapeControlData<T>
//                   : T extends ListControlDefinition
//                     ? GetItemListControlData<T>
//                     : T extends LinkControlDefinition
//                       ? LinkControlData
//                       : T extends RichTextControlDefinition
//                         ? IndexSignatureHack<RichTextControlData>
//                         : T extends RichTextV2ControlDefinition
//                           ? RichTextV2ControlData
//                           : T extends StyleControlDefinition
//                             ? StyleControlData
//                             : T extends StyleV2ControlDefinition
//                               ? StyleV2ControlData
//                               : T extends TypographyControlDefinition
//                                 ? TypographyControlData
//                                 : never

// export type GetItemLabelListControlItemData<T extends ListControlDefinition> = {
//   id: string
//   type?: T['config']['type']['type']
//   value: GetItemLabelControlData<T['config']['type']>
// }

// export type GetItemLabelShapeControlData<
//   T extends ShapeControlDefinition = ShapeControlDefinition,
// > = {
//   [K in keyof T['config']['type']]?: GetItemLabelControlData<
//     T['config']['type'][K]
//   >
// }

// export type GetItemListControlData<
//   T extends ListControlDefinition = ListControlDefinition,
// > = GetItemLabelListControlItemData<T>[]

// type ListControlConfig<T extends ControlDefinition = ControlDefinition> = {
//   type: T
//   label?: string
//   getItemLabel?(item: GetItemLabelControlData<T> | undefined): string
// }

// export type ListControlDefinition<
//   C extends ListControlConfig = ListControlConfig,
// > = {
//   type: typeof ListControlType
//   config: C
// }

import { match } from 'ts-pattern'
import { z } from 'zod'

import {
  ControlDataTypeKey,
  colorDataSchema,
  type ColorData,
  type CopyContext,
  type ValueType,
  Swatch,
  ControlDataType,
} from '../common'

import { controlTraitsRegistry } from '../registry'
import {
  type ResourceResolver,
  type ResolvableValue,
} from '../resource-resolver'

import {
  type VersionedControlDefinition,
  type ControlTraits,
  type ParseResult,
} from '../traits'

import { DefaultControlInstance, type Send } from '../control-instance'
import { WithAssociatedTypes } from '../utils/associated-types'

export const List = controlTraitsRegistry.add(
  (() => {
    const type = 'makeswift::controls::list' as const
    // const v1DataType = 'list::v1' as const
    // const version = 1 as const

    // const dataSignature = {
    //   v1: { [ControlDataTypeKey]: v1DataType },
    // } as const

    // const dataSchema = z.union([
    //   colorDataSchema,
    //   colorDataSchema.and(
    //     z.object({
    //       [ControlDataTypeKey]: z.literal(v1DataType),
    //     }),
    //   ),
    // ])

    // type ControlData = z.infer<typeof dataSchema>

    type ItemData<ItemDef extends ControlDefinition> = {
      id: string
      type?: ItemDef['type']
      value: ControlDataType<ItemDef>
    }

    // export type ListControlData<
    //   T extends ListControlDefinition = ListControlDefinition,
    // > = ListControlItemData<T>[]

    type Config<
      Def extends VersionedControlDefinition = VersionedControlDefinition,
    > = {
      type: Def
      label?: string
      getItemLabel?(item: ControlDataType<Def> | undefined): string
    }

    type ControlDefinition<C extends Config = Config> = {
      type: typeof type
      config: C
    } & WithAssociatedTypes<{
      ControlType: typeof type
      Config: Config
      ValueType: ValueType
      ResolvedValueType: ResolvedValueType
    }>
    //   typeof type,
    //   C,
    //   ColorData,
    //   typeof version,
    //   undefined extends C['defaultValue'] ? string | undefined : string
    // >

    // export type ListControlDefinition<
    //   C extends ListControlConfig = ListControlConfig,
    // > = {
    //   type: typeof ListControlType
    //   config: C
    // }

    const ctor = <C extends Config>(config?: C): ControlDefinition<C> => ({
      type,
      config: config ?? ({} as C),
      version,
    })

    ctor.controlType = type
    ctor.dataSignature = dataSignature

    ctor.safeParse = (
      data: unknown | undefined,
    ): ParseResult<ControlData | undefined> => {
      const result = dataSchema.optional().safeParse(data)
      return result.success
        ? { success: true, data: result.data }
        : { success: false, error: result.error.flatten().formErrors[0] }
    }

    ctor.fromData = (
      data: ControlData | undefined,
      _definition: ControlDefinition,
    ) => {
      return match(data)
        .with(dataSignature.v1, ({ swatchId, alpha }) => ({ swatchId, alpha }))
        .otherwise((value) => value ?? null)
    }

    ctor.toData = (
      value: ColorData,
      definition: ControlDefinition,
    ): ControlData => {
      return match('version' in definition ? definition.version : undefined)
        .with(version, () => ({
          ...dataSignature.v1,
          ...value,
        }))
        .with(undefined, () => value)
        .exhaustive()
    }

    ctor.copyData = (
      data: ControlData | undefined,
      { replacementContext }: CopyContext,
    ): ControlData | undefined => {
      if (data == null) return data

      const replaceSwatchId = (swatchId: string) =>
        replacementContext.swatchIds.get(swatchId) ?? swatchId

      return match(data)
        .with(dataSignature.v1, (val) => ({
          ...val,
          swatchId: replaceSwatchId(val.swatchId),
        }))
        .otherwise((val) => ({
          ...val,
          swatchId: replaceSwatchId(val.swatchId),
        }))
    }

    ctor.getSwatchIds = (data: ControlData | undefined): string[] =>
      data?.swatchId == null ? [] : [data.swatchId]

    ctor.resolveValue = (
      value: ValueType<ControlDefinition>,
      definition: ControlDefinition,
      resolver: ResourceResolver,
    ): ResolvableValue<string | undefined> => {
      const { swatchId, alpha } = value ?? {}
      return resolver
        .resolveSwatch(swatchId)
        .map((swatch) =>
          swatchToColorString(swatch, alpha, definition.config.defaultValue),
        )
    }

    ctor.createInstance = (send: Send) => new DefaultControlInstance(send)

    return ctor as typeof ctor &
      ControlTraits<typeof type, ControlData, ControlDefinition>
  })(),
)

// export function List<
//   T extends ControlDefinition,
//   C extends ListControlConfig<T>,
// >(config: C & { type: T }): ListControlDefinition<C> {
//   return { type: ListControlType, config }
// }

// export const ListControlMessageType = {
//   LIST_CONTROL_ITEM_CONTROL_MESSAGE:
//     'makeswift::controls::list::message::item-control-message',
// } as const

// type ListControlItemControlMessage = {
//   type: typeof ListControlMessageType.LIST_CONTROL_ITEM_CONTROL_MESSAGE
//   payload: { message: PropControllerMessage; itemId: string }
// }

// export type ListControlMessage = ListControlItemControlMessage

// export class ListControl<
//   T extends ListControlDefinition = ListControlDefinition,
// > extends PropController<ListControlMessage> {
//   controls: Map<string, AnyPropController>
//   descriptor: ListControlDefinition
//   send: Send<ListControlMessage>

//   constructor(send: Send<ListControlMessage>, descriptor: T) {
//     super(send)

//     this.descriptor = descriptor
//     this.send = send

//     this.controls = new Map<string, AnyPropController>()
//   }

//   setItemsControl = (value: ListControlData<T> | undefined) => {
//     const controls = new Map<string, AnyPropController>()

//     if (value == null) return

//     const shouldUpdate = () => {
//       // If the length is different, should update
//       if (value.length !== this.controls.size) return true
//       // If this.controls does not have an itemId, should update
//       if (!value.every(({ id }) => this.controls.has(id))) return true

//       return false
//     }

//     if (!shouldUpdate()) return this.controls

//     value.forEach((item) => {
//       const control = createPropController(
//         this.descriptor.config.type,
//         (message) =>
//           this.send({
//             type: ListControlMessageType.LIST_CONTROL_ITEM_CONTROL_MESSAGE,
//             payload: { message, itemId: item.id },
//           }),
//       )

//       controls.set(item.id, control)
//     })

//     this.controls = controls

//     return this.controls
//   }

//   recv = (message: ListControlMessage) => {
//     switch (message.type) {
//       case ListControlMessageType.LIST_CONTROL_ITEM_CONTROL_MESSAGE: {
//         const control = this.controls.get(message.payload.itemId)

//         if (control == null) return

//         // TODO: We're casting the type here as the arg0 type for control.recv is never
//         const recv = control.recv as (arg0: PropControllerMessage) => void

//         recv(message.payload.message)
//       }
//     }
//   }
// }

// export function copyListData(
//   definition: ListControlDefinition,
//   value: ListControlData | undefined,
//   context: CopyContext,
// ): ListControlData | undefined {
//   if (value == null) return value

//   return (
//     value &&
//     value.map((item) => ({
//       ...item,
//       value: controlCopy(definition.config.type, item.value, context),
//     }))
//   )
// }

// function introspectListData<T>(
//   definition: ListControlDefinition,
//   value: ListControlData | undefined,
//   func: (definition: ControlDefinition, data: Data) => T[],
// ): T[] {
//   if (value == null) return []

//   return value.flatMap((item) => func(definition.config.type, item.value))
// }

// export function getListElementChildren(
//   definition: ListControlDefinition,
//   value: ListControlData,
// ) {
//   return introspectListData(definition, value, getElementChildren)
// }

// export function getListSwatchIds(
//   definition: ListControlDefinition,
//   value: ListControlData,
// ) {
//   return introspectListData(definition, value, getSwatchIds)
// }

// export function getListFileIds(
//   definition: ListControlDefinition,
//   value: ListControlData,
// ) {
//   return introspectListData(definition, value, getFileIds)
// }

// export function getListTypographyIds(
//   definition: ListControlDefinition,
//   value: ListControlData,
// ) {
//   return introspectListData(definition, value, getTypographyIds)
// }

// export function getListPageIds(
//   definition: ListControlDefinition,
//   value: ListControlData,
// ) {
//   return introspectListData(definition, value, getPageIds)
// }

// export function getListTranslatableData(
//   definition: ListControlDefinition,
//   data: ListControlData,
// ) {
//   return Object.fromEntries(
//     data.map((item) => [
//       item.id,
//       getTranslatableData(definition.config.type, item.value),
//     ]),
//   )
// }

// export type ListControlTranslationDto = Record<string, ListControlData>

// export function mergeListTranslatedData(
//   definition: ListControlDefinition,
//   data: ListControlData,
//   translatedData: ListControlTranslationDto,
//   context: MergeTranslatableDataContext,
// ) {
//   return data.map((item) => {
//     return {
//       ...item,
//       value: mergeTranslatedData(
//         definition.config.type,
//         item.value,
//         translatedData[item.id],
//         context,
//       ),
//     }
//   })
// }
