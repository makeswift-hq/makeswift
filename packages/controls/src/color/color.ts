import { match } from 'ts-pattern'
import { z } from 'zod'

import { ControlDataTypeKey, colorDataSchema } from '../common'
import { type CopyContext } from '../context'
import {
  type ResourceResolver,
  type ValueSubscription,
} from '../resource-resolver'

import {
  DefaultControlInstance,
  ControlInstance,
  type SendType,
} from '../control-instance'

import {
  ControlDefinition,
  safeParse,
  serialize,
  type ParseResult,
} from '../control-definition'

import { swatchToColorString } from './conversion'

type Config = z.infer<typeof Definition.schema.relaxed.config>

type DataSchemaType<C extends Config> = undefined extends C['defaultValue']
  ? typeof Definition.schema.relaxed
  : typeof Definition.schema.strict

type DataType<C extends Config> = z.infer<DataSchemaType<C>['data']>
type ValueType<C extends Config> = z.infer<DataSchemaType<C>['value']>
type ResolvedValueType<C extends Config> = z.infer<
  DataSchemaType<C>['resolvedValue']
>

class Definition<C extends Config = Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  private static readonly v1DataType = 'color::v1' as const
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const

  static readonly type = 'makeswift::controls::color' as const

  static get schema() {
    const type = z.literal(this.type)
    const version = z.literal(1).optional()
    const value = colorDataSchema

    const schemas = <R>(resolvedValue: z.ZodType<R>) => {
      const data = value.and(
        z.object({
          [ControlDataTypeKey]: z.literal(this.v1DataType).optional(),
        }),
      )

      const config = z.object({
        label: z.string().optional(),
        labelOrientation: z
          .union([z.literal('horizontal'), z.literal('vertical')])
          .optional(),
        defaultValue: resolvedValue,
        hideAlpha: z.boolean().optional(),
      })

      const definition = z.object({
        type,
        config,
        version,
      })

      return { type, value, resolvedValue, data, config, version, definition }
    }

    return {
      version,
      relaxed: schemas(z.string().optional()),
      strict: schemas(z.string()),
    }
  }

  static deserialize(data: unknown): Definition {
    const { config, version } = this.schema.relaxed.definition.parse(data)
    return new Definition(config, version)
  }

  constructor(
    config: C,
    readonly version: z.infer<typeof Definition.schema.version>,
  ) {
    super(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    return (
      this.config.defaultValue === undefined
        ? Definition.schema.relaxed
        : Definition.schema.strict
    ) as DataSchemaType<C>
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    const inputSchema = this.schema.data.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, ({ swatchId, alpha }) => ({
        swatchId,
        alpha,
      }))
      .otherwise((value) => value)
  }

  toData(value: ValueType<C>): DataType<C> {
    return match(this.version)
      .with(1, () => ({
        ...Definition.dataSignature.v1,
        ...value,
      }))
      .with(undefined, () => value)
      .exhaustive()
  }

  copyData(
    data: DataType<C> | undefined,
    { replacementContext }: CopyContext,
  ): DataType<C> | undefined {
    if (data == null) return data

    const replaceSwatchId = (swatchId: string) =>
      replacementContext.swatchIds.get(swatchId) ?? swatchId

    const inputSchema = this.schema.data.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, (val) => ({
        ...val,
        swatchId: replaceSwatchId(val.swatchId),
      }))
      .otherwise((val) => ({
        ...val,
        swatchId: replaceSwatchId(val.swatchId),
      }))
  }

  resolveValue(
    value: ValueType<C>,
    resolver: ResourceResolver,
  ): ValueSubscription<ResolvedValueType<C>> {
    const swatchSubscription = resolver.subscribeSwatch(value?.swatchId)

    if (value?.swatchId != null && swatchSubscription.readValue() == null) {
      resolver.fetchSwatch(value.swatchId).catch(console.error)
    }

    return {
      readValue: () =>
        swatchToColorString(
          swatchSubscription.readValue(),
          value?.alpha ?? 1,
          this.config.defaultValue,
        ),
      subscribe: swatchSubscription.subscribe,
    }
  }

  createInstance(send: SendType<ControlInstance>): ControlInstance {
    return new DefaultControlInstance(send)
  }

  serialize(): [unknown, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
      version: this.version,
    })
  }

  getSwatchIds(data: DataType<C> | undefined): string[] {
    return data?.swatchId == null ? [] : [data.swatchId]
  }
}

export const Color = <C extends Config>(config?: C) =>
  new (class Color extends Definition<C> {})(config ?? ({} as C), 1)

export { Definition as ColorDefinition }
