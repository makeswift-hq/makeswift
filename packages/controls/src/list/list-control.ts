import { type DataType } from '../control-definition'
import {
  ControlInstance,
  type ControlMessage,
  type Send,
} from '../control-instance'

import { type ListDefinition } from './list'

type Message = {
  type: typeof ListControl.messageType.ITEM_CONTROL_MESSAGE
  payload: { message: ControlMessage; itemId: string }
}

export class ListControl<
  Def extends ListDefinition = ListDefinition,
> extends ControlInstance<Message> {
  static readonly messageType = {
    ITEM_CONTROL_MESSAGE:
      'makeswift::controls::list::message::item-control-message',
  } as const

  private itemControls: Map<string, ControlInstance> = new Map()

  constructor(
    private readonly definition: Def,
    send: Send<Message>,
  ) {
    super(send)
  }

  setItemsControl = (data: DataType<Def> | undefined) => {
    if (data == null) return

    const shouldUpdate = () => {
      // If the length is different, should update
      if (data.length !== this.itemControls.size) return true
      // If this.controls does not have an itemId, should update
      if (!data.every(({ id }) => this.itemControls.has(id))) return true

      return false
    }

    if (!shouldUpdate()) return this.itemControls

    return (this.itemControls = new Map(
      data.map(({ id }) => [id, this.createItemControl(id)]),
    ))
  }

  recv = (message: Message) => {
    switch (message.type) {
      case ListControl.messageType.ITEM_CONTROL_MESSAGE: {
        const control = this.itemControls.get(message.payload.itemId)
        control?.recv(message.payload.message)
      }
    }
  }

  createItemControl = (id: string) => {
    return this.definition.itemDef.createInstance((message) =>
      this.send({
        type: ListControl.messageType.ITEM_CONTROL_MESSAGE,
        payload: { message, itemId: id },
      }),
    )
  }
}
