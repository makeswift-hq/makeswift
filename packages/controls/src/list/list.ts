import { z } from 'zod'

// import { MappedResolvableValue } from '../resource-resolver'
import { type CopyContext } from '../context'
// import {
//   type ResourceResolver,
// } from '../resource-resolver'

import { type SendType } from '../control-instance'

import {
  ControlDefinition,
  safeParse,
  serialize,
  type Schema,
  type ParseResult,
  type DataType as DataType_,
  type ValueType as ValueType_,
  type ResolvedValueType as ResolvedValueType_,
} from '../control-definition'

import { notNil, keyNotNil } from '../utils/functional'

import { ListControl } from './list-control'

type Config<Item extends ControlDefinition = ControlDefinition> = {
  type: Item
  label?: string
  getItemLabel?(item: DataType_<Item> | undefined): string
}

type ItemType<C extends Config> = C extends Config<infer Item> ? Item : never
type DataType<C extends Config> = {
  id: string
  type?: string
  value: DataType_<ItemType<C>>
}[]

type ValueType<C extends Config> = ValueType_<ItemType<C>>[]
type ResolvedValueType<C extends Config> = ResolvedValueType_<ItemType<C>>[]

class Definition<C extends Config = Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>,
  ListControl<Definition<C>>
> {
  static readonly type = 'makeswift::controls::list' as const

  constructor(readonly config: C) {
    super(config)
  }

  get itemDef() {
    return this.config.type
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    const type = z.literal(Definition.type)
    const item = this.itemDef.schema

    const data = z.array(
      z.object({
        id: z.string(),
        type: item.type.optional(),
        value: item.data,
      }),
    ) as Schema<DataType<C> | undefined>

    const value = z.array(item.value)
    const resolvedValue = z.array(item.resolvedValue).optional()

    const config = z.object({
      type: item.definition,
      label: z.string().optional(),
      getItemLabel: z.function().optional(),
    })

    const definition = z.object({
      type,
      config,
    })

    return { type, data, value, resolvedValue, config, definition }
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    return data?.map(({ value }) => this.itemDef.fromData(value)).filter(notNil)
  }

  toData(value: ValueType<C>): DataType<C> {
    return value.map((value, index) => ({
      id: `${index}`,
      type: this.itemDef.controlType,
      value: this.itemDef.toData(value),
    }))
  }

  copyData(
    data: DataType<C> | undefined,
    context: CopyContext,
  ): DataType<C> | undefined {
    return data
      ?.map((item) => ({
        ...item,
        value: this.itemDef.copyData(item.value, context),
      }))
      .filter((data) => keyNotNil(data, 'value'))
  }

  // resolveValue(
  //   value: ValueType<C>,
  //   resolver: ResourceResolver,
  // ): ResolvableValue<ResolvedValueType<C>> {
  //   const resolved = value?.map((value) =>
  //     this.itemDef.resolveValue(value, resolver),
  //   )

  //   return new MappedResolvableValue(
  //     () => resolved?.map((r) => r.readValue()).filter(notNil),
  //     (_onUpdate: () => void) => () => {}, // FIXME: implement
  //     (value: ResolvedValueType<C>) => value,
  //   )
  // }

  createInstance(
    send: SendType<ListControl<Definition<C>>>,
  ): ListControl<Definition<C>> {
    return new ListControl(this, send)
  }

  serialize(): [unknown, Transferable[]] {
    // FIXME: implement
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  getSwatchIds(data: DataType<C> | undefined): string[] {
    return data?.flatMap((item) => this.itemDef.getSwatchIds(item.value)) ?? []
  }

  // introspect<R extends Resource>(
  //   data: DataType<C> | undefined,
  //   r: R,
  // ): ReturnType<R['introspect']> {
  //   return data?.flatMap((item) => this.itemDef.introspect(item.value, r)) ?? []
  // }
}

export const List = <C extends Config>(config: C) =>
  new (class List extends Definition<C> {})(config)

export { Definition as ListDefinition }
