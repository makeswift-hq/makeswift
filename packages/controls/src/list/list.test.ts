import { List } from './list'
import { Color } from '../color'
// import {
//   type ValueType,
//   type ResolvedValueType,
//   DataType,
// } from '../control-definition'

describe('List', () => {
  describe('constructor', () => {
    test('list of colors', () => {
      const list = List({
        type: Color({ defaultValue: 'red', labelOrientation: 'horizontal' }),
        label: 'Color list',
      })
      expect(list).toMatchSnapshot()
      //   type D = DataType<typeof list>
      //   type V = ValueType<typeof list>
      //   type R = ResolvedValueType<typeof list>
      //   const r: R = [['#ff0000', undefined], ['#00ff00']]
      // colorList: List({ label: 'Color list', type: Color({ defaultValue: 'blue' }) }),
    })
  })
})
