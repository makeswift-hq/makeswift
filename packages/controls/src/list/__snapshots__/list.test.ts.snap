// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List constructor list of colors 1`] = `
List {
  "config": {
    "label": "Color list",
    "type": Color {
      "config": {
        "defaultValue": "red",
        "labelOrientation": "horizontal",
      },
      "version": 1,
    },
  },
}
`;
